helm-app-operator/pkg/helm/helm.go

```diff
@@ -254,30 +258,36 @@ func (c installer) ReconcileRelease(r *unstructured.Unstructured) (*unstructured
 	}
 
 	var updatedRelease *release.Release
-	latestRelease, err := c.storageBackend.Deployed(releaseName(r))
+	latestRelease, err := c.storageBackend.Deployed(releaseName)
 	if err != nil || latestRelease == nil {
-		updatedRelease, err = c.installRelease(r, tiller, chart, config)
+		// If there's no deployed release, attempt a tiller install.
+		updatedRelease, err = c.installRelease(tiller, r.GetNamespace(), releaseName, chart, config)
 		if err != nil {
 			return r, needsUpdate, fmt.Errorf("install error: %s", err)
 		}
 		needsUpdate = true
 		logrus.Infof("Installed release for %s release=%s", ResourceString(r), updatedRelease.GetName())
+
+	} else if status.Release == nil {
```

Comment:
@joelanford I'm wondering if this scenario is possible:
- There is no deployed release, so we install the release `c.installRelease()`
- We [fail to update](https://github.com/operator-framework/helm-app-operator-kit/blob/master/helm-app-operator/pkg/helm/controller/reconcile.go#L97-L101) the `status.Release` 
- Retry the reconcile loop and this time `c.storageBackend.Deployed()` returns `latestRelease, nil`.
- `status.Release==nil` from our last failed status update and we return an error thinking this release is not owned by the CR(even though it actually is).

Reply:
Yeah, that definitely sounds possible. Off the top of my head, I can think of a few things we can try:
* If we fail to update, try to rollback the release to whatever the previous `status.Release` was, or uninstall if it was nil. Depending on the error, the rollback may also fail.
* Rather than using `c.storageBackend.Deployed()` as the latestRelease, use `status.Release`. 
* Keep an in-memory map of resource UUID to release name.
* See if there's a way to store the resource UUID in the release (maybe with the chart values), such that `c.storageBackend.Deployed()` would "know" its owner.

Each of these may cause other interesting failure scenarios though, so I'd have to experiment to know whether any of these are feasible.

Any other ideas?

Reply:
> See if there's a way to store the resource UUID in the release (maybe with the chart values), such that c.storageBackend.Deployed() would "know" its owner.

Turns out it's pretty easy to add an arbitrary config value to the release, which gets saved with the release in the storage backend. It seems like this is probably more robust than the other options. The latest commit changes the ownership checks to use this method instead.